### PR TITLE
Waggle Dance Core Changes For Pluggable ThriftHiveMetastore.IFaces

### LINKED ISSUE #93: Enable custom ThriftHiveMetastore.IFace implementations
Enable configurable pluggable implementations of the ThriftHiveMetastore.IFace interface on a per (primary or federated) metastore basis. 

Example of how the configuration could look like
```
primary-meta-store:
  access-control-type: READ_ONLY
  database-prefix: ''
  name: primary
  closeable-iface: com.company.foo.bar.MyImpl
  #com.company.foo.bar.MyImpl is a custom 
  #implementation of the CloseableThriftHiveMetastoreIface  that can forward
  #requests to some service in the background and adapt them to the Hive interface 
  #e.g. federate Glue onto hive, federate BigQuery onto hive etc by using this
  #interface as an adapter from your data store to the 
  #CloseableThriftHiveMetastoreIface  interface 

federated-meta-stores:
- access-control-type: READ_ONLY
  database-prefix: waggle_test_
  mapped-databases: []
  name: secondary
  remote-meta-store-uris: thrift://waggle-dance:48869
  #No closeable-iface implementation is specified so it defaults to Thrift 

  #Thrift Hive will be federated onto your custom implementation

```

### PR DESCRIPTION
**Pull request by courtsVII**
resolves #93  

Changes to enable configurable pluggable implementations of the ThriftHiveMetastore.IFace interface on a per (primary or federated) metastore basis. Adds configuration parameter `closeable-iface` which is added to the metastore configuration in your yaml and specifies a fully qualified class path to a implementation of the `CloseableThriftHiveMetastoreIface` interface (the WaggleDance wrapper over `ThriftHiveMetastore.IFace` `Tunnelable` and `Closeable`. 

The class for your implementation must be on the WaggleDance classpath, this can be achieved by either adding your project as a maven dependency in waggle-dance-core, or adding the jar with your plugin to /opt/waggle-dance/lib/ in your waggle-dance installation.

Example config with these changes federating custom implementation onto prefixed Hive metastore:

```
primary-meta-store:
  access-control-type: READ_ONLY
  database-prefix: ''
  name: primary
  configuration-properties:
        com.hotels.bdp.waggledance.client.impl: com.company.foo.bar.MyImpl
  #com.com.company.foo.bar.MyImpl is a custom 
  #implementation of the CloseableThriftHiveMetastoreIface  that can forward
  #requests to some service in the background and adapt them to the Hive interface 
  #e.g. federate Glue onto hive, federate BigQuery onto hive etc by using this
  #interface as an adapter from your data store to the 
  #CloseableThriftHiveMetastoreIface  interface 

federated-meta-stores:
- access-control-type: READ_ONLY
  database-prefix: waggle_test_
  mapped-databases: []
  name: secondary
  remote-meta-store-uris: thrift://waggle-dance:48869

```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -139,18 +139,22 @@ The table below describes all the available configuration values for Waggle Danc
 | Property                                              | Required | Description |
 |:----|:----:|:----|
 | `primary-meta-store`                                  | No       | Primary MetaStore config. Can be empty but it is advised to configure it. |
-| `primary-meta-store.remote-meta-store-uris`           | Yes      | Thrift URIs of the federated read-only metastore. |
```
Comment (courtsVII) [inline README.md]: Not all metastore have a remote-meta-store-uri (e.g. glue), this is a hive thing and shouldn't be in our AbstractMetaStore definition, it should be optional

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIface.java [hunk 2]
```diff
@@ -1,28 +0,0 @@
...
- */
-package com.hotels.bdp.waggledance.client;
-
-import java.io.Closeable;
-
-import org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore.Iface;
-
-import com.hotels.hcommon.ssh.Tunnelable;
-
-public interface CloseableThriftHiveMetastoreIface extends Iface, Closeable, Tunnelable {
```
Comment (courtsVII) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/CloseableThriftHiveMetastoreIface.java]: I moved this to `hcommon-hive-metastore` so we can decouple implementations of the interface for pluggable components from `WaggleDanceCore`

File: pom.xml [hunk 3]
```diff
@@ -43,6 +44,7 @@
     <jdk.version>1.7</jdk.version>
     <aspectj.version>1.8.9</aspectj.version>
     <beeju.version>1.2.0</beeju.version>
+    <hcommon-hive-metastore.version>1.1.1-SNAPSHOT</hcommon-hive-metastore.version>
```
Comment (courtsVII) [inline pom.xml]: Bump from Snapshot version once hcommon-hive-metastore 1.1.1 is released

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 4]
```diff
@@ -102,20 +108,38 @@ private void reconnectIfDisconnected() {
...
-        hiveConf, name, reconnectionRetries);
-    return (CloseableThriftHiveMetastoreIface) Proxy.newProxyInstance(getClass().getClassLoader(), INTERFACES,
-        reconnectingHandler);
-  }
+    CloseableThriftHiveMetastoreIface closeableThriftHiveMetastoreIface = null;
+    final String closeableIFaceImpl = hiveConf.get(WaggleDanceHiveConfVars.CLOSEABLE_IFACE_IMPL.varname);
+    if (isNotEmpty(closeableIFaceImpl)) {
+      try {
+        Class<?> clazz = Class.forName(closeableIFaceImpl);
+        Constructor<?> constructor = clazz.getConstructor(HiveConf.class);
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java]: Where we document the property that allows people to customise the client implementation we need to clearly point out what interface it needs to implement and that it also needs to provide a matching constructor.

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/WaggleDanceHiveConfVars.java [hunk 5]
```diff
@@ -23,7 +23,8 @@
   SSH_PRIVATE_KEYS("com.hotels.bdp.waggledance.client.ssh.private_keys"),
   SSH_KNOWN_HOSTS("com.hotels.bdp.waggledance.client.ssh.known_hosts"),
   SSH_SESSION_TIMEOUT("com.hotels.bdp.waggledance.client.ssh.session_timeout"),
-  SSH_STRICT_HOST_KEY_CHECKING("com.hotels.bdp.waggledance.client.ssh.strict_host_key_checking");
+  SSH_STRICT_HOST_KEY_CHECKING("com.hotels.bdp.waggledance.client.ssh.strict_host_key_checking"),
+  CLOSEABLE_IFACE_IMPL("com.hotels.bdp.waggledance.client.impl");
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/WaggleDanceHiveConfVars.java]: CLOSEABLE_IFACE_IMPL -> CUSTOM_HIVE_CLIENT_IMPLEMENTATION ? (more descriptive as to what this is, the variable in the above class that is pulled out of here should be renamed accordingly)

File: waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java [hunk 6]
```diff
@@ -102,20 +108,38 @@ private void reconnectIfDisconnected() {
...
   @Override
   public CloseableThriftHiveMetastoreIface newInstance(HiveConf hiveConf, String name, int reconnectionRetries) {
-    ReconnectingMetastoreClientInvocationHandler reconnectingHandler = new ReconnectingMetastoreClientInvocationHandler(
-        hiveConf, name, reconnectionRetries);
-    return (CloseableThriftHiveMetastoreIface) Proxy.newProxyInstance(getClass().getClassLoader(), INTERFACES,
-        reconnectingHandler);
-  }
+    CloseableThriftHiveMetastoreIface closeableThriftHiveMetastoreIface = null;
+    final String closeableIFaceImpl = hiveConf.get(WaggleDanceHiveConfVars.CLOSEABLE_IFACE_IMPL.varname);
+    if (isNotEmpty(closeableIFaceImpl)) {
```
Comment (massdosage) [inline waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/client/DefaultMetaStoreClientFactory.java]: Could do with some tests of this logic where we pass in a test class that is a bare bones client implementation and ensure this gets constructed properly. Then again but missing the required constructor to ensure the exception gets rethrown. Also pass in non-existing class name.

**Comment by massdosage** [comment]
> Close this now perhaps?

Yeah, I guess the pressing need for this isn't there right now, we would revisit the approach if someone asked again in future anyway.

**Pull request closed without merging**